File: songclone/api/events.py
"""SSE event schemas and broadcasting infrastructure."""

import asyncio
from datetime import datetime
from enum import Enum
from typing import Any, Literal, Optional, Union
from uuid import uuid4

from pydantic import BaseModel, Field


class TimestampedEvent(BaseModel):
    """Base class for all SSE events with automatic timestamps."""
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PhaseType(str, Enum):
    ANALYSIS = "analysis"
    ITERATION = "iteration"


class EventStatus(str, Enum):
    STARTED = "started"
    COMPLETE = "complete"


class StepType(str, Enum):
    PLANNING = "planning"
    EXECUTION = "execution"
    EVALUATION = "evaluation"


class LogLevel(str, Enum):
    INFO = "info"
    WARN = "warn"
    ERROR = "error"


class CompleteReason(str, Enum):
    THRESHOLD_REACHED = "threshold_reached"
    MAX_ITERATIONS = "max_iterations"
    CANCELLED = "cancelled"


class PhaseEvent(TimestampedEvent):
    type: Literal["phase"] = "phase"
    phase: PhaseType
    status: EventStatus
    data: Optional[Any] = None


class IterationEvent(TimestampedEvent):
    type: Literal["iteration"] = "iteration"
    number: int
    status: EventStatus


class StepEvent(TimestampedEvent):
    type: Literal["step"] = "step"
    step: StepType
    status: EventStatus
    data: Optional[Any] = None


class LogEvent(TimestampedEvent):
    type: Literal["log"] = "log"
    level: LogLevel
    message: str


class AudioEvent(TimestampedEvent):
    type: Literal["audio"] = "audio"
    iteration: int
    path: str


class HumanActionEvent(TimestampedEvent):
    type: Literal["human_action_required"] = "human_action_required"
    description: str = Field(..., description="What action is needed")
    reason: str = Field(..., description="Why automation failed")
    steps: list[str] = Field(..., description="Step-by-step instructions")


class PausedEvent(TimestampedEvent):
    type: Literal["paused"] = "paused"
    reason: str = Field(..., description="Why the session was paused")


class CompleteEvent(TimestampedEvent):
    type: Literal["complete"] = "complete"
    reason: CompleteReason
    iterations: int = Field(..., description="Total iterations completed")


class ErrorEvent(TimestampedEvent):
    type: Literal["error"] = "error"
    message: str
    recoverable: bool


class ReaperOperationType(str, Enum):
    CREATE_PROJECT = "create_project"
    BATCH_CREATE_TRACKS = "batch_create_tracks"
    BATCH_INSERT_MIDI = "batch_insert_midi"
    BATCH_SET_FX = "batch_set_fx"
    BATCH_SET_LEVELS = "batch_set_levels"
    RENDER_AUDIO = "render_audio"


class ReaperOperationEvent(TimestampedEvent):
    type: Literal["reaper_operation"] = "reaper_operation"
    operation: ReaperOperationType
    status: EventStatus
    details: Optional[dict[str, Any]] = None
    error: Optional[str] = None


class ToolCallEvent(TimestampedEvent):
    """ADK tool call event."""
    type: Literal["tool_call"] = "tool_call"
    tool: str
    call_id: str = Field(default_factory=lambda: str(uuid4())[:8])
    args: Optional[dict[str, Any]] = None


class ToolResponseEvent(TimestampedEvent):
    """ADK tool response event."""
    type: Literal["tool_response"] = "tool_response"
    tool: str
    call_id: str
    status: str
    result: Optional[dict[str, Any]] = None
    duration_ms: Optional[int] = None


class AgentResponseEvent(TimestampedEvent):
    """ADK agent final response."""
    type: Literal["agent_response"] = "agent_response"
    content: str


class AgentThinkingEvent(TimestampedEvent):
    """ADK agent thinking/reasoning event."""
    type: Literal["agent_thinking"] = "agent_thinking"
    content: str
    agent: str = "orchestrator"


class TraceUrlEvent(TimestampedEvent):
    """Langfuse trace URL for observability."""
    type: Literal["trace_url"] = "trace_url"
    url: str


SSEEvent = Union[
    PhaseEvent,
    IterationEvent,
    StepEvent,
    LogEvent,
    AudioEvent,
    HumanActionEvent,
    PausedEvent,
    CompleteEvent,
    ErrorEvent,
    ReaperOperationEvent,
    ToolCallEvent,
    ToolResponseEvent,
    AgentResponseEvent,
    AgentThinkingEvent,
    TraceUrlEvent,
]


class EventEmitter:
    """Async event emitter for SSE broadcasting to connected clients."""

    def __init__(self) -> None:
        self._subscribers: dict[str, list[asyncio.Queue[SSEEvent]]] = {}

    def subscribe(self, session_id: str) -> asyncio.Queue[SSEEvent]:
        """Subscribe to events for a session. Returns a queue to receive events."""
        if session_id not in self._subscribers:
            self._subscribers[session_id] = []
        queue: asyncio.Queue[SSEEvent] = asyncio.Queue()
        self._subscribers[session_id].append(queue)
        return queue

    def unsubscribe(self, session_id: str, queue: asyncio.Queue[SSEEvent]) -> None:
        """Unsubscribe from events for a session."""
        if session_id in self._subscribers:
            try:
                self._subscribers[session_id].remove(queue)
            except ValueError:
                pass
            if not self._subscribers[session_id]:
                del self._subscribers[session_id]

    async def emit(self, session_id: str, event: SSEEvent) -> None:
        """Emit an event to all subscribers of a session."""
        if session_id in self._subscribers:
            for queue in self._subscribers[session_id]:
                await queue.put(event)

    def emit_sync(self, session_id: str, event: SSEEvent) -> None:
        """Synchronous version for use in non-async contexts."""
        if session_id in self._subscribers:
            for queue in self._subscribers[session_id]:
                queue.put_nowait(event)


event_emitter = EventEmitter()

File: songclone/web/src/types/events.ts
/**
 * SSE event types matching Python Pydantic schemas.
 * Per constitution X: Events MUST be typed (TypeScript interfaces matching Python schemas).
 */

export type PhaseType = 'analysis' | 'iteration';
export type EventStatus = 'started' | 'complete';
export type StepType = 'planning' | 'execution' | 'evaluation';
export type LogLevel = 'info' | 'warn' | 'error';
export type CompleteReason = 'threshold_reached' | 'max_iterations' | 'cancelled';
export type ReaperOperationType =
  | 'create_project'
  | 'batch_create_tracks'
  | 'batch_insert_midi'
  | 'batch_set_fx'
  | 'batch_set_levels'
  | 'render_audio';

interface BaseEvent {
  timestamp: string;
}

export interface PhaseEvent extends BaseEvent {
  type: 'phase';
  phase: PhaseType;
  status: EventStatus;
  data?: unknown;
}

export interface IterationEvent extends BaseEvent {
  type: 'iteration';
  number: number;
  status: EventStatus;
}

export interface StepEvent extends BaseEvent {
  type: 'step';
  step: StepType;
  status: EventStatus;
  data?: unknown;
}

export interface LogEvent extends BaseEvent {
  type: 'log';
  level: LogLevel;
  message: string;
}

export interface AudioEvent extends BaseEvent {
  type: 'audio';
  iteration: number;
  path: string;
}

export interface HumanActionEvent extends BaseEvent {
  type: 'human_action_required';
  description: string;
  reason: string;
  steps: string[];
}

export interface PausedEvent extends BaseEvent {
  type: 'paused';
  reason: string;
}

export interface CompleteEvent extends BaseEvent {
  type: 'complete';
  reason: CompleteReason;
  iterations: number;
}

export interface ErrorEvent extends BaseEvent {
  type: 'error';
  message: string;
  recoverable: boolean;
}

export interface ReaperOperationEvent extends BaseEvent {
  type: 'reaper_operation';
  operation: ReaperOperationType;
  status: EventStatus;
  details?: Record<string, unknown>;
  error?: string;
}

export interface ToolCallEvent extends BaseEvent {
  type: 'tool_call';
  tool: string;
  call_id: string;
  args?: Record<string, unknown>;
}

export interface ToolResponseEvent extends BaseEvent {
  type: 'tool_response';
  tool: string;
  call_id: string;
  status: string;
  result?: Record<string, unknown>;
  duration_ms?: number;
}

export interface AgentResponseEvent extends BaseEvent {
  type: 'agent_response';
  content: string;
}

export interface AgentThinkingEvent extends BaseEvent {
  type: 'agent_thinking';
  content: string;
  agent: string;
}

export type SSEEvent =
  | PhaseEvent
  | IterationEvent
  | StepEvent
  | LogEvent
  | AudioEvent
  | HumanActionEvent
  | PausedEvent
  | CompleteEvent
  | ErrorEvent
  | ReaperOperationEvent
  | ToolCallEvent
  | ToolResponseEvent
  | AgentResponseEvent
  | AgentThinkingEvent;

// Type guard functions for narrowing event types
export function isPhaseEvent(event: SSEEvent): event is PhaseEvent {
  return event.type === 'phase';
}

export function isIterationEvent(event: SSEEvent): event is IterationEvent {
  return event.type === 'iteration';
}

export function isStepEvent(event: SSEEvent): event is StepEvent {
  return event.type === 'step';
}

export function isLogEvent(event: SSEEvent): event is LogEvent {
  return event.type === 'log';
}

export function isAudioEvent(event: SSEEvent): event is AudioEvent {
  return event.type === 'audio';
}

export function isHumanActionEvent(event: SSEEvent): event is HumanActionEvent {
  return event.type === 'human_action_required';
}

export function isPausedEvent(event: SSEEvent): event is PausedEvent {
  return event.type === 'paused';
}

export function isCompleteEvent(event: SSEEvent): event is CompleteEvent {
  return event.type === 'complete';
}

export function isErrorEvent(event: SSEEvent): event is ErrorEvent {
  return event.type === 'error';
}

export function isReaperOperationEvent(event: SSEEvent): event is ReaperOperationEvent {
  return event.type === 'reaper_operation';
}

export function isToolCallEvent(event: SSEEvent): event is ToolCallEvent {
  return event.type === 'tool_call';
}

export function isToolResponseEvent(event: SSEEvent): event is ToolResponseEvent {
  return event.type === 'tool_response';
}

export function isAgentResponseEvent(event: SSEEvent): event is AgentResponseEvent {
  return event.type === 'agent_response';
}

export function isAgentThinkingEvent(event: SSEEvent): event is AgentThinkingEvent {
  return event.type === 'agent_thinking';
}

// Session and iteration types matching API schemas

export interface Scores {
  timing: number;
  harmony: number;
  melody: number;
  instruments: number;
  mix: number;
  overall: number;
}

export interface FeedbackItem {
  priority: number;
  category: 'timing' | 'harmony' | 'melody' | 'instruments' | 'mix' | 'other';
  issue: string;
  suggestion: string;
}

export interface EvaluationResult {
  scores: Scores;
  total_score: number;
  max_score: number;
  feedback: FeedbackItem[];
  stop_early: boolean;
  stop_reason: string | null;
  evaluation_method: 'ai' | 'mfcc_fallback';
}

export interface InstrumentConfig {
  vst: string;
  preset: string | null;
}

export interface MixSettings {
  volume_db: number;
  pan: number;
  mute: boolean;
}

export interface FXConfig {
  plugin: string;
  preset: string | null;
  params: Record<string, unknown> | null;
}

export interface TrackPlan {
  name: string;
  role: 'vocals' | 'drums' | 'bass' | 'other' | 'master';
  instrument: InstrumentConfig;
  midi_source: string;
  midi_transform: { transpose: number; velocity_scale: number } | null;
  mix: MixSettings;
  fx: FXConfig[];
}

export interface ExecutionPlan {
  reasoning: string;
  tracks: TrackPlan[];
  master_fx: FXConfig[];
}

export interface Iteration {
  number: number;
  plan: ExecutionPlan;
  render_path: string | null;
  evaluation: EvaluationResult | null;
  started_at: string;
  completed_at: string | null;
  status: 'planning' | 'executing' | 'evaluating' | 'completed' | 'failed';
}

export interface Session {
  id: string;
  status: 'uploading' | 'analyzing' | 'iterating' | 'paused' | 'completed' | 'failed' | 'cancelled';
  original_audio_path: string | null;
  iterations: Iteration[];
  current_iteration: number;
  max_iterations: number;
  quality_threshold: number;
  error: string | null;
}
